File: examples/dotnet/client/GameManager.cs
// Copyright (c) 2026 solar@heliacal.net
// SPDX-License-Identifier: MIT

using System;
using System.Diagnostics;
using RdplibExample.Common;

namespace RdplibExample.Client
{
    internal sealed class GameManager
    {
        private static readonly TimeSpan PositionUpdateInterval = TimeSpan.FromSeconds(1);
        private static readonly TimeSpan ProfileSaveInterval = TimeSpan.FromSeconds(6);
        private static readonly TimeSpan GameDuration = TimeSpan.FromSeconds(30);

        private readonly Stopwatch _gameTimer;
        private bool _started;
        private TimeSpan _nextPositionUpdate;
        private TimeSpan _nextProfileSave;
        private uint _positionSequence;
        private uint _saveSequence;
        private float _x;
        private float _y;
        private float _z;
        private float _heading;

        internal GameManager()
        {
            _gameTimer = new Stopwatch();
            _z = 0.0f;
        }

        internal bool ProcessGame(ServerManager server)
        {
            if (!_started)
            {
                _started = true;
                _nextPositionUpdate = TimeSpan.Zero;
                _nextProfileSave = ProfileSaveInterval;
                _gameTimer.Start();
                Console.WriteLine("game simulation started");
            }

            TimeSpan now = _gameTimer.Elapsed;
            if (now >= GameDuration)
            {
                if (!server.SendLogout(LogoutReason.ExampleComplete))
                    return false;

                Console.WriteLine("requested logout with reason {0}", (uint)LogoutReason.ExampleComplete);
                return true;
            }

            // ClientPositionUpdate is an example unreliable gameplay message.
            if (now >= _nextPositionUpdate)
            {
                ++_positionSequence;
                _x += 1.0f;
                _y += 0.5f;
                _heading += 8.0f;

                ClientPositionUpdate update = new ClientPositionUpdate();
                update.Sequence = _positionSequence;
                update.X = _x;
                update.Y = _y;
                update.Z = _z;
                update.Heading = _heading;

                if (!server.SendClientPositionUpdate(update))
                    return false;

                Console.WriteLine("sent unreliable position update {0}", update.Sequence);
                _nextPositionUpdate = now + PositionUpdateInterval;
            }

            // SaveProfile is an example reliable gameplay message.
            if (now >= _nextProfileSave)
            {
                SaveProfile profile = new SaveProfile();
                profile.Sequence = ++_saveSequence;
                profile.SecondsInGame = (uint)now.TotalSeconds;
                profile.X = _x;
                profile.Y = _y;
                profile.Z = _z;
                profile.Heading = _heading;

                if (!server.SendSaveProfile(profile))
                    return false;

                Console.WriteLine("sent reliable profile save {0}", profile.Sequence);
                _nextProfileSave = now + ProfileSaveInterval;
            }

            return true;
        }
    }
}

File: examples/cpp/client/game_manager.cpp
// Copyright (c) 2026 solar
// SPDX-License-Identifier: MIT

#include "game_manager.h"

#include "application_protocol.h"
#include "server_manager.h"

#include <iostream>

GameManager::GameManager()
    : m_started(false),
      m_position_sequence(0),
      m_save_sequence(0),
      m_x(0.0f),
      m_y(0.0f),
      m_z(0.0f),
      m_heading(0.0f)
{
}

bool GameManager::ProcessGame(ServerManager &server)
{
    auto now = std::chrono::steady_clock::now();
    if (!m_started)
    {
        m_started = true;
        m_started_at = now;
        m_next_position_update = now;
        m_next_profile_save = now + ProfileSaveInterval;
        m_logout_time = now + GameDuration;
        std::cout << "game simulation started\n";
    }

    if (now >= m_logout_time)
    {
        if (!server.SendLogout(LogoutReason::ExampleComplete))
            return false;

        std::cout << "requested logout with reason " << static_cast<std::uint32_t>(LogoutReason::ExampleComplete) << '\n';
        return true;
    }

    // ClientPositionUpdate - an example unreliable gameplay message sent to the server
    if (now >= m_next_position_update)
    {
        ++m_position_sequence;
        m_x += 1.0f;
        m_y += 0.5f;
        m_heading += 8.0f;

        ClientPositionUpdate update{};
        update.sequence = m_position_sequence;
        update.x = m_x;
        update.y = m_y;
        update.z = m_z;
        update.heading = m_heading;

        if (!server.SendClientPositionUpdate(update))
            return false;

        std::cout << "sent unreliable position update " << update.sequence << '\n';
        m_next_position_update = now + PositionUpdateInterval;
    }

    // SaveProfile - an example reliable gameplay message sent to the server
    if (now >= m_next_profile_save)
    {
        SaveProfile profile{};
        profile.sequence = ++m_save_sequence;
        profile.seconds_in_game = static_cast<std::uint32_t>(
            std::chrono::duration_cast<std::chrono::seconds>(now - m_started_at).count()
        );
        profile.x = m_x;
        profile.y = m_y;
        profile.z = m_z;
        profile.heading = m_heading;

        if (!server.SendSaveProfile(profile))
            return false;

        std::cout << "sent reliable profile save " << profile.sequence << '\n';
        m_next_profile_save = now + ProfileSaveInterval;
    }

    return true;
}
